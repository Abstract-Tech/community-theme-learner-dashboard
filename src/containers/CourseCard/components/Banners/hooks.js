import { useIntl } from '@edx/frontend-platform/i18n';
import { useCardValues } from 'hooks';
import { selectors } from 'data/redux';

const { cardData } = selectors;

export const useCourseBannerData = ({ courseNumber }) => ({
  courseData: useCardValues(courseNumber, {
    isVerified: cardData.isVerified,
    isCourseRunActive: cardData.isCourseRunActive,
    canUpgrade: cardData.canUpgrade,
    isAuditAccessExpired: cardData.isAuditAccessExpired,
    courseWebsite: cardData.courseWebsite,
  }),
  formatMessage: useIntl().formatMessage,
});

export default {
  useCourseBannerData,
};
