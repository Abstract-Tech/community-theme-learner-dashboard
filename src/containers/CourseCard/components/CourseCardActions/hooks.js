import { useDispatch } from 'react-redux';

import { Locked } from '@edx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import { hooks as appHooks } from 'data/redux';
import messages from './messages';

export const useCardActionData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const dispatch = useDispatch();
  const {
    canUpgrade,
    coursewareAccess,
    hasStarted,
    isAudit,
    isAuditAccessExpired,
    isVerified,
  } = appHooks.useCardEnrollmentData(cardId);
  const { isArchived } = appHooks.useCardCourseRunData(cardId);
  const {
    canChange,
    hasSessions,
    isEntitlement,
    isExpired,
    isFulfilled,
  } = appHooks.useCardEntitlementData(cardId);
  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(dispatch, cardId);

  const hasAccess = coursewareAccess.isStaff || !(
    coursewareAccess.hasUnmetPrereqs
    || coursewareAccess.isTooEarly
    || isArchived);

  const selectSessionButton = {
    children: formatMessage(messages.selectSession),
    onClick: openSessionModal,
    disabled: !hasAccess || (!canChange || !hasSessions),
  };

  const viewCourseButton = {
    children: formatMessage(messages.viewCourse),
    disabled: !hasAccess || (isEntitlement && isExpired),
  };

  const beginCourseButton = {
    children: formatMessage(messages.beginCourse),
    disabled: !hasAccess,
  };

  const resumeButton = {
    children: formatMessage(messages.resume),
    disabled: isAudit && isAuditAccessExpired,
  };

  const upgradeButton = {
    iconBefore: Locked,
    variant: 'outline-primary',
    children: formatMessage(messages.upgrade),
    disabled: !canUpgrade,
  };

  let primary;
  if (isEntitlement) {
    primary = isFulfilled ? selectSessionButton : viewCourseButton;
  } else if (!hasStarted) {
    primary = beginCourseButton;
  } else if (isArchived) {
    primary = viewCourseButton;
  } else {
    primary = resumeButton;
  }

  const secondary = (isEntitlement || !isVerified) ? null : upgradeButton;
  return { primary, secondary };
};

export default useCardActionData;
