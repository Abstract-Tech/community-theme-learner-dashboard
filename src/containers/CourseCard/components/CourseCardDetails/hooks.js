import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

import * as module from './hooks';
import messages from './messages';

export const useAccessMessage = ({ cardId }) => {
  const { formatMessage, formatDate } = useIntl();
  const enrollment = appHooks.useCardEnrollmentData(cardId);
  const courseRun = appHooks.useCardCourseRunData(cardId);
  if (!courseRun.isStarted) {
    return formatMessage(messages.courseStarts, { startDate: courseRun.startDate });
  }
  if (enrollment.isEnrolled) {
    if (enrollment.isAudit) {
      const {
        accessExpirationDate,
        isAuditAccessExpired,
      } = enrollment;
      return formatMessage(
        isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
        { accessExpirationDate: formatDate(accessExpirationDate) },
      );
    }
    const { isArchived, endDate } = courseRun;
    return formatMessage(
      isArchived ? messages.courseEnded : messages.courseEnds,
      { endDate: formatDate(endDate) },
    );
  }
  return null;
};

export const useCardDetailsData = ({ dispatch, cardId }) => {
  const { formatMessage } = useIntl();
  const providerName = appHooks.useCardProviderData(cardId).name;
  const { courseNumber } = appHooks.useCardCourseData(cardId);
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = appHooks.useCardEntitlementsData(cardId);

  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(dispatch, cardId);

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: module.useAccessMessage({ cardId }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    formatMessage,
    courseNumber,
  };
};

export default useCardDetailsData;
