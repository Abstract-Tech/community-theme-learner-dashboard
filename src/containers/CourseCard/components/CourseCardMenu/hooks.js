import React from 'react';
import { StrictDict } from 'utils';
import * as module from './hooks';

export const state = StrictDict({
  isUnenrollConfirmVisible: (val) => React.useState(val),
  isEmailSettingsVisible: (val) => React.useState(val),
});

export const unenrollModalHooks = () => {
  const [isVisible, setIsVisible] = module.state.isUnenrollConfirmVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const emailSettingsModalHooks = () => {
  const [isVisible, setIsVisible] = module.state.isEmailSettingsVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const menuHooks = () => {
  const unenrollModal = module.unenrollModalHooks();
  const emailSettingsModal = module.emailSettingsModalHooks();
  return {
    emailSettingsModal,
    unenrollModal,
  };
};

export default menuHooks;
