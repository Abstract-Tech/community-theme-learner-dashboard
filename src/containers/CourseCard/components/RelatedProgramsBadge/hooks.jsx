import React from 'react';
import { useIntl } from '@edx/frontend-platform/i18n';

import { StrictDict } from 'utils';
import { hooks as appHooks } from 'data/redux';

import messages from './messages';
import * as module from './hooks';

export const state = StrictDict({
  isOpen: (val) => React.useState(val), // eslint-disable-line
});

export const useRelatedProgramsBadgeData = ({ cardId }) => {
  const [isOpen, setIsOpen] = module.state.isOpen(false);
  const { formatMessage } = useIntl();
  const numPrograms = appHooks.useCardRelatedProgramsData(cardId).length;
  let programsMessage = '';
  if (numPrograms) {
    programsMessage = formatMessage(
      numPrograms === 1 ? messages.badgeLabelSingular : messages.badgeLabelPlural,
      { numPrograms },
    );
  }

  return {
    numPrograms,
    programsMessage,
    isOpen,
    openModal: React.useCallback(() => setIsOpen(true), [setIsOpen]),
    closeModal: React.useCallback(() => setIsOpen(false), [setIsOpen]),
  };
};

export default useRelatedProgramsBadgeData;
