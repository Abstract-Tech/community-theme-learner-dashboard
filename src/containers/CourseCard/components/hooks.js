import { reduxHooks } from 'hooks';
import { EXECUTIVE_EDUCATION_COURSE_MODES } from '../../../data/constants/course';

export const useActionDisabledState = (cardId) => {
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const {
    canUpgrade, hasAccess, isAudit, isAuditAccessExpired, mode,
  } = reduxHooks.useCardEnrollmentData(cardId);
  const {
    isEntitlement, isFulfilled, canChange, hasSessions,
  } = reduxHooks.useCardEntitlementData(cardId);

  const { resumeUrl, homeUrl, upgradeUrl } = reduxHooks.useCardCourseRunData(cardId);

  const disableBeginCourse = !homeUrl || (isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired));
  const disableResumeCourse = !resumeUrl || (isMasquerading || !hasAccess || (isAudit && isAuditAccessExpired));
  const disableViewCourse = !hasAccess || (isAudit && isAuditAccessExpired);
  const disableSelectSession = !isEntitlement || isMasquerading || !hasAccess || (!canChange || !hasSessions);
  const disableUpgradeCourse = !upgradeUrl || (isMasquerading && !canUpgrade);

  const disableCourseTitle = (isEntitlement && !isFulfilled) || disableViewCourse;

  const isExecutiveEd2uCourse = (EXECUTIVE_EDUCATION_COURSE_MODES.includes(mode));

  return {
    disableBeginCourse,
    disableResumeCourse,
    disableViewCourse,
    disableUpgradeCourse,
    disableSelectSession,
    disableCourseTitle,
    isExecutiveEd2uCourse,
  };
};

export default useActionDisabledState;
