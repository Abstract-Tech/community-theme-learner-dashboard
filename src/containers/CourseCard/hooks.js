import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

import * as module from './hooks';
import messages from './messages';

export const useAccessMessage = ({ courseNumber }) => {
  const { formatMessage, formatDate } = useIntl();
  const {
    accessExpirationDate,
    isAudit,
    isAuditAccessExpired,
  } = appHooks.useCardEnrollmentData(courseNumber);
  const { isArchived, endDate } = appHooks.useCardCourseRunData(courseNumber);

  if (isAudit) {
    return formatMessage(
      isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
      { accessExpirationDate: formatDate(accessExpirationDate) },
    );
  }

  return formatMessage(
    isArchived ? messages.courseEnded : messages.courseEnds,
    { endDate: formatDate(endDate) },
  );
};

export const useCardData = ({ courseNumber }) => {
  const { formatMessage } = useIntl();
  const { title, bannerUrl } = appHooks.useCardCourseData(courseNumber);
  const providerName = appHooks.useCardProviderData(courseNumber).name;

  return {
    title,
    bannerUrl,
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: module.useAccessMessage({ courseNumber }),
    formatMessage,
  };
};

export default useCardData;
