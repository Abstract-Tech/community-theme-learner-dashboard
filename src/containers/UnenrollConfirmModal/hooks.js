import React from 'react';

import { StrictDict } from 'utils';
import { thunkActions } from 'data/redux';

import * as module from './hooks';

export const state = StrictDict({
  confirmed: (val) => React.useState(val),
  customReason: (val) => React.useState(val),
  isSkipped: (val) => React.useState(val),
  selectedReason: (val) => React.useState(val),
  submittedReason: (val) => React.useState(val),
});

export const modalStates = StrictDict({
  confirm: 'confirm',
  reason: 'reason',
  finished: 'finished',
});

export const unenrollReasons = () => {
  const [selectedReason, setSelectedReason] = module.state.selectedReason(null);
  const [submittedReason, setSubmittedReason] = module.state.submittedReason(null);
  const [isSkipped, setIsSkipped] = module.state.isSkipped(false);
  const [customOption, setCustomOption] = module.state.customReason('');

  return {
    clear: React.useCallback(() => {
      setSelectedReason(null);
      setSubmittedReason(null);
      setCustomOption('');
      setIsSkipped(false);
    }, []),
    customOption: {
      value: customOption,
      onChange: React.useCallback((e) => setCustomOption(e.target.value), []),
    },
    isSkipped,
    isSubmitted: submittedReason !== null || isSkipped,
    selected: selectedReason,
    selectOption: React.useCallback((e) => setSelectedReason(e.target.value), []),
    skip: React.useCallback(() => setIsSkipped(true), [isSkipped]),
    submit: React.useCallback(() => {
      if (selectedReason === 'custom') {
        setSubmittedReason(customOption);
      } else {
        setSubmittedReason(selectedReason);
      }
    }, [customOption, selectedReason]),
    value: submittedReason,
  };
};

export const modalHooks = ({ closeModal, dispatch }) => {
  const [isConfirmed, setIsConfirmed] = module.state.confirmed(false);

  const confirm = React.useCallback(() => setIsConfirmed(true), []);

  const reason = unenrollReasons();
  const close = () => {
    closeModal();
    setIsConfirmed(false);
    reason.clear();
  };

  let modalState;
  if (isConfirmed) {
    modalState = reason.isSubmitted ? modalStates.finished : modalState.reason;
  } else {
    modalState = modalStates.confirm;
  }

  const closeAndRefresh = React.useCallback(() => {
    dispatch(thunkActions.app.refreshList());
    close();
  }, []);

  return {
    isConfirmed,
    confirm,
    reason,
    closeAndRefresh,
    close,
    modalState,
  };
};

export default modalHooks;
