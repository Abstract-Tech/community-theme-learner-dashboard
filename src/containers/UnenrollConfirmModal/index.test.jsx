import React from 'react';
import { shallow } from 'enzyme';
import { useDispatch } from 'react-redux';

import { UnenrollConfirmModal } from '.';

import * as hooks from './hooks';

jest.mock('./components/ConfirmPane', () => 'ConfirmPane');
jest.mock('./components/ReasonPane', () => 'ReasonPane');
jest.mock('./components/FinishedPane', () => 'FinishedPane');

jest.mock('./hooks', () => ({
  __esModule: true,
  modalStates: jest.requireActual('./hooks').modalStates,
  modalHooks: jest.fn(),
}));

describe('UnenrollConfirmModal component', () => {
  const dispatch = useDispatch();
  const hookProps = {
    confirm: jest.fn().mockName('hooks.confirm'),
    reason: {
      isSkipped: false,
      reasonProps: 'other',
    },
    close: jest.fn().mockName('hooks.close'),
    closeAndRefresh: jest.fn().mockName('hooks.closeAndRefresh'),
    modalState: hooks.modalStates.confirm,
  };
  const closeModal = jest.fn().mockName('props.closeModal');
  const show = true;
  test('hooks called with dispatch and closeModal props', () => {
    hooks.modalHooks.mockReturnValueOnce(hookProps);
    shallow(<UnenrollConfirmModal {...{ closeModal, show }} />);
    expect(hooks.modalHooks).toHaveBeenCalledWith({ dispatch, closeModal });
  });
  test('snapshot: modalStates.confirm', () => {
    hooks.modalHooks.mockReturnValueOnce(hookProps);
    expect(shallow(<UnenrollConfirmModal {...{ closeModal, show }} />)).toMatchSnapshot();
  });
  test('snapshot: modalStates.finished, reason given', () => {
    hooks.modalHooks.mockReturnValueOnce({ ...hookProps, modalState: hooks.modalStates.finished });
    expect(shallow(<UnenrollConfirmModal {...{ closeModal, show }} />)).toMatchSnapshot();
  });
  test('snapshot: modalStates.finished, reason skipped', () => {
    hooks.modalHooks.mockReturnValueOnce({
      ...hookProps,
      modalState: hooks.modalStates.finished,
      isSkipped: true,
    });
    expect(shallow(<UnenrollConfirmModal {...{ closeModal, show }} />)).toMatchSnapshot();
  });
  test('snapshot: modalStates.reason', () => {
    hooks.modalHooks.mockReturnValueOnce({ ...hookProps, modalState: hooks.modalStates.reason });
    expect(shallow(<UnenrollConfirmModal {...{ closeModal, show }} />)).toMatchSnapshot();
  });
});
