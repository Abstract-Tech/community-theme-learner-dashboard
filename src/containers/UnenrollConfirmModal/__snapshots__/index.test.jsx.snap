// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UnenrollConfirmModal component snapshot: modalStates.confirm 1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  title=""
>
  <div
    className="bg-white p-3 rounded shadow"
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    <ConfirmPane
      handleClose={[MockFunction hooks.close]}
      handleConfirm={[MockFunction hooks.confirm]}
    />
  </div>
</ModalDialog>
`;

exports[`UnenrollConfirmModal component snapshot: modalStates.finished, reason given 1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  title=""
>
  <div
    className="bg-white p-3 rounded shadow"
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    <FinishedPane
      gaveReason={true}
      handleClose={[MockFunction hooks.closeAndRefresh]}
    />
  </div>
</ModalDialog>
`;

exports[`UnenrollConfirmModal component snapshot: modalStates.finished, reason skipped 1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  title=""
>
  <div
    className="bg-white p-3 rounded shadow"
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    <FinishedPane
      gaveReason={true}
      handleClose={[MockFunction hooks.closeAndRefresh]}
    />
  </div>
</ModalDialog>
`;

exports[`UnenrollConfirmModal component snapshot: modalStates.reason 1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  title=""
>
  <div
    className="bg-white p-3 rounded shadow"
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    <ReasonPane
      reason={
        Object {
          "isSkipped": false,
          "reasonProps": "other",
        }
      }
    />
  </div>
</ModalDialog>
`;
